conformance/utils/kubernetes/apply.go

```diff
@@ -41,45 +42,95 @@ import (
 // Applier prepares manifests depending on the available options and applies
 // them to the Kubernetes cluster.
 type Applier struct {
-	NamespaceLabels map[string]string
-	// ValidUniqueListenerPorts maps each listener port of each Gateway in the
-	// manifests to a valid, unique port. There must be as many
-	// ValidUniqueListenerPorts as there are listeners in the set of manifests.
-	// For example, given two Gateways, each with 2 listeners, there should be
-	// four ValidUniqueListenerPorts.
-	// If empty or nil, ports are not modified.
-	ValidUniqueListenerPorts []v1alpha2.PortNumber
+	namespaceLabels map[string]string
+	// availablePorts holds ports that can still be assigned
+	availablePorts PortSet
+	// assignedPorts tracks ports used for Gateway objects
+	assignedPorts map[types.NamespacedName]PortSet
+	portsLock     sync.Mutex
```

Comment:
Is this lock intended to guard assignedPorts, availablePorts or both?

Reply:
It not only guards the map from concurrent writes but creates a critical section for freeing and allocating ports every time we have a Gateway.